File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/core/bounded_execution.rb
# frozen_string_literal: true

require "securerandom"
require_relative "../audit/event"
require_relative "../audit/null_sink"
require_relative "pressure_config"
require_relative "pressure_engine"

module IAIso
  module Core
    # High-level execution facade. Composes a PressureEngine with an
    # audit sink and lifecycle management.
    #
    # Use BoundedExecution.run with a block for automatic cleanup, or
    # BoundedExecution.start + manual #close for explicit control.
    class BoundedExecution
      attr_reader :engine

      def initialize(execution_id: nil, config: PressureConfig.defaults,
                     audit_sink: nil, clock: nil, timestamp_clock: nil)
        execution_id = "exec-#{SecureRandom.hex(6)}" if execution_id.nil? || execution_id.empty?
        @audit_sink = audit_sink || Audit::NullSink.instance
        @ts_clock = timestamp_clock || clock || WallClock.instance
        @engine = PressureEngine.new(
          config: config,
          options: EngineOptions.new(
            execution_id: execution_id,
            audit_sink: @audit_sink,
            clock: clock || WallClock.instance,
            timestamp_clock: @ts_clock,
          ),
        )
        @closed = false
        @lock = Monitor.new
      end

      # Construct + return a started execution. The caller MUST #close it.
      def self.start(**opts)
        new(**opts)
      end

      # Run a block inside a bounded execution; closes on exit.
      def self.run(**opts)
        exec = new(**opts)
        errored = false
        begin
          yield exec
        rescue StandardError
          errored = true
          raise
        ensure
          exec.send(:close_with, errored: errored)
        end
      end

      def execution_id = @engine.execution_id
      def snapshot     = @engine.snapshot

      def record_tokens(tokens, tag: nil)
        @engine.step(StepInput.build(tokens: tokens, tag: tag))
      end

      def record_tool_call(name, tokens: 0)
        @engine.step(StepInput.build(tokens: tokens, tool_calls: 1, tag: name))
      end

      def record_step(work)
        @engine.step(work)
      end

      # Pre-check the engine state without advancing it.
      def check
        case @engine.lifecycle
        when Lifecycle::LOCKED then StepOutcome::LOCKED
        when Lifecycle::ESCALATED then StepOutcome::ESCALATED
        else StepOutcome::OK
        end
      end

      def reset = @engine.reset

      # Close the execution, emitting execution.closed. Idempotent.
      def close
        close_with(errored: false)
      end

      private

      def close_with(errored:)
        @lock.synchronize do
          return if @closed
          @closed = true
        end
        snap = @engine.snapshot
        @audit_sink.emit(Audit::Event.new(
          execution_id: @engine.execution_id,
          kind: "execution.closed",
          timestamp: @ts_clock.now,
          data: {
            "final_pressure" => snap.pressure,
            "final_lifecycle" => snap.lifecycle,
            "exception" => errored ? "error" : nil,
          },
        ))
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoCore/BoundedExecution.swift
import Foundation
import IAIsoAudit

/// High-level execution facade. Composes a `PressureEngine` with an
/// audit sink and lifecycle management.
///
/// Use `BoundedExecution.run(_:body:)` for closure-style with automatic
/// cleanup, or `start(_:)` + manual `close()` for explicit control.
public final class BoundedExecution: @unchecked Sendable {

    public struct Options: Sendable {
        public var executionId: String?
        public var config: PressureConfig
        public var auditSink: Sink
        public var clock: Clock
        public var timestampClock: Clock?

        public init(
            executionId: String? = nil,
            config: PressureConfig = .defaults,
            auditSink: Sink = NullSink.shared,
            clock: Clock = WallClock.shared,
            timestampClock: Clock? = nil
        ) {
            self.executionId = executionId
            self.config = config
            self.auditSink = auditSink
            self.clock = clock
            self.timestampClock = timestampClock
        }
    }

    public let engine: PressureEngine
    private let auditSink: Sink
    private let timestampClock: Clock
    private let lock = NSLock()
    private var _closed = false

    private init(_ options: Options) throws {
        let execId = options.executionId.flatMap { $0.isEmpty ? nil : $0 }
            ?? "exec-\(Self.randomHex(6))"
        let tsClock = options.timestampClock ?? options.clock
        self.auditSink = options.auditSink
        self.timestampClock = tsClock
        self.engine = try PressureEngine(
            config: options.config,
            options: EngineOptions(
                executionId: execId,
                auditSink: options.auditSink,
                clock: options.clock,
                timestampClock: tsClock))
    }

    /// Construct a `BoundedExecution`. The caller MUST `close()` it.
    public static func start(_ options: Options = Options()) throws -> BoundedExecution {
        return try BoundedExecution(options)
    }

    /// Run a closure inside a bounded execution; closes on exit.
    public static func run(
        _ options: Options = Options(),
        body: (BoundedExecution) throws -> Void
    ) throws {
        let exec = try start(options)
        var errored = false
        do {
            try body(exec)
        } catch {
            errored = true
            exec.closeWith(errored: true)
            throw error
        }
        exec.closeWith(errored: errored)
    }

    public func snapshot() -> PressureSnapshot { engine.snapshot() }

    /// Account for tokens with an optional tag.
    @discardableResult
    public func recordTokens(_ tokens: Int64, tag: String? = nil) -> StepOutcome {
        return engine.step(StepInput(tokens: tokens, tag: tag))
    }

    /// Account for a single tool invocation.
    @discardableResult
    public func recordToolCall(_ name: String, tokens: Int64 = 0) -> StepOutcome {
        return engine.step(StepInput(tokens: tokens, toolCalls: 1, tag: name))
    }

    /// General step accounting.
    @discardableResult
    public func recordStep(_ work: StepInput) -> StepOutcome {
        return engine.step(work)
    }

    /// Pre-check the engine state without advancing it.
    public func check() -> StepOutcome {
        switch engine.lifecycle {
        case .locked: return .locked
        case .escalated: return .escalated
        default: return .ok
        }
    }

    @discardableResult
    public func reset() -> PressureSnapshot { engine.reset() }

    /// Close the execution, emitting `execution.closed`. Idempotent.
    public func close() {
        closeWith(errored: false)
    }

    private func closeWith(errored: Bool) {
        lock.lock()
        if _closed {
            lock.unlock()
            return
        }
        _closed = true
        lock.unlock()
        let snap = engine.snapshot()
        auditSink.emit(Event(
            executionId: engine.executionId,
            kind: "execution.closed",
            timestamp: timestampClock.now(),
            data: [
                "final_pressure": .double(snap.pressure),
                "final_lifecycle": .string(snap.lifecycle.rawValue),
                "exception": errored ? .string("error") : .null,
            ]))
    }

    deinit {
        // Defensive: guarantee close is emitted even if user forgot.
        closeWith(errored: false)
    }

    private static func randomHex(_ bytes: Int) -> String {
        var data = Data(count: bytes)
        for i in 0..<bytes {
            data[i] = UInt8.random(in: 0...UInt8.max)
        }
        return data.map { String(format: "%02x", $0) }.joined()
    }
}
